lib/sshutils/networking/networking.go

```diff
@@ -0,0 +1,399 @@
+	listenerFD, err := p.sendRequest(ctx, Request{
+		Operation:  NetworkingOperationListenX11,
+		X11Request: req,
+	})
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	listener, err := net.FileListener(listenerFD.file)
+	if err != nil {
+		listenerFD.Close()
+		return nil, trace.Wrap(err)
+	}
+
+	return &childListener{
+		Listener:            listener,
+		childNetworkingFile: listenerFD,
+	}, nil
+}
+
+// childNetworkingFile is a networking file corresponding to a networking
+// resource opened in the child process, with a lifetime associated with
+// the per-request connection.
+type childNetworkingFile struct {
+	conn *net.UnixConn
+	file *os.File
+}
+
+func (f *childNetworkingFile) Close() error {
+	return trace.NewAggregate(f.conn.Close(), f.file.Close())
+}
+
+// childConn is a net.Conn backed by a connection in the child process.
+type childConn struct {
+	childNetworkingResource *childNetworkingFile
+	net.Conn
+}
+
+// Close the local file connection and the child networking resources behind it.
+func (c *childConn) Close() error {
+	return trace.NewAggregate(c.Conn.Close(), c.childNetworkingResource.Close())
+}
+
+// childListener is a net.Listener backed by a listener in the child process.
+type childListener struct {
+	childNetworkingFile *childNetworkingFile
+	net.Listener
+}
+
+// Close the local file listener and the child networking resources behind it.
+func (l *childListener) Close() error {
+	return trace.NewAggregate(l.Listener.Close(), l.childNetworkingFile.Close())
+}
+
+// sendRequest sends a networking request to the networking process and waits
+// for a file corresponding to an open network connection or listener.
+func (p *Process) sendRequest(ctx context.Context, req Request) (*childNetworkingFile, error) {
+	slog.DebugContext(ctx, "Sending networking request to child process", "request", req)
+
+	ctx, cancel := context.WithTimeout(ctx, 30*time.Second)
+	defer cancel()
+
+	jsonReq, err := json.Marshal(req)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	// We can use a stream instead of a datagram because we only expect bytes
+	// or a file descriptor in response, not both.
+	requestConn, remoteConn, err := uds.NewSocketpair(uds.SocketTypeStream)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	remoteFD, err := remoteConn.File()
+	remoteConn.Close()
+	if err != nil {
+		requestConn.Close()
+		return nil, trace.Wrap(err)
+	}
+
+	if _, _, err = uds.WriteWithFDs(p.conn, jsonReq, []*os.File{remoteFD}); err != nil {
+		requestConn.Close()
+		remoteFD.Close()
+		return nil, trace.Wrap(err)
+	}
+	remoteFD.Close()
+
+	// Read in another goroutine so we can cancel it if the networking process stops or the context closes.
+	fileCh := make(chan *os.File)
+	errC := make(chan error)
+	go func() {
+		defer close(fileCh)
+		file, err := readResponse(requestConn)
+		if err != nil {
+			errC <- err
+		} else {
+			fileCh <- file
```

Comment:
Case in point to avoid goroutines if not strictly necessary: this will block forever (leaking the goroutine stack and the entry in the scheduler) if an object or an error is being pushed into the channels and the context or p.done unblock, so nothing is reading in any of those two channels anymore.